cmd/juju/charmhub/printinfo.go

```diff
@@ -0,0 +1,535 @@
+
+	// establish the indent of the whole block
+	var err error
+	for len(text) > width && err == nil {
+		// find a good place to chop the text
+		idx := runesLastIndexSpace(text[:width+1])
+		if idx < 0 {
+			// there's no whitespace; just chop at line width
+			idx = width
+		}
+		_, err = fmt.Fprint(out, indent, string(text[:idx]), "\n")
+		// prune any remaining whitespace before the start of the next line
+		for idx < len(text) && unicode.IsSpace(text[idx]) {
+			idx++
+		}
+		text = text[idx:]
+		width += delta
+		indent = indent2
+		delta = 0
+	}
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprint(out, indent, string(text), "\n")
+	return err
+}
+
+// runesLastIndexSpace returns the index of the last whitespace rune
+// in the text. If the text has no whitespace, returns -1.
+func runesLastIndexSpace(text []rune) int {
+	for i := len(text) - 1; i >= 0; i-- {
+		if unicode.IsSpace(text[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// printDescr formats a given string (typically a snap description)
+// in a user friendly way.
+//
+// The rules are (intentionally) very simple:
+// - trim trailing whitespace
+// - word wrap at "max" chars preserving line indent
+// - keep \n intact and break there
+func printDescr(w io.Writer, descr string, termWidth int) error {
+	var err error
+	descr = strings.TrimRightFunc(descr, unicode.IsSpace)
+	for _, line := range strings.Split(descr, "\n") {
+		err = wrapLine(w, []rune(line), "  ", termWidth)
+		if err != nil {
+			break
+		}
+	}
+	return err
+}
+
+// wrapLine wraps a line, assumed to be part of a block-style yaml
+// string, to fit into termWidth, preserving the line's indent, and
+// writes it out prepending padding to each line.
+func wrapLine(out io.Writer, text []rune, pad string, termWidth int) error {
+	// discard any trailing whitespace
+	text = runesTrimRightSpace(text)
+	// establish the indent of the whole block
+	idx := 0
+	for idx < len(text) && unicode.IsSpace(text[idx]) {
+		idx++
+	}
+	indent := pad + string(text[:idx])
+	text = text[idx:]
+	if len(indent) > termWidth/2 {
+		// If indent is too big there's not enough space for the actual
+		// text, in the pathological case the indent can even be bigger
+		// than the terminal which leads to lp:1828425.
+		// Rather than let that happen, give up.
+		indent = pad + "  "
+	}
+	return wrapGeneric(out, text, indent, indent, termWidth)
+}
+
+// runesTrimRightSpace returns text, with any trailing whitespace dropped.
+func runesTrimRightSpace(text []rune) []rune {
+	j := len(text)
+	for j > 0 && unicode.IsSpace(text[j-1]) {
+		j--
+	}
+	return text[:j]
+}
+
+func termWidth(w io.Writer) int {
+	var width int
+	if f, ok := w.(*os.File); ok {
+		width, _, _ = terminal.GetSize(int(f.Fd()))
+	}
+
+	if width <= 0 {
+		width = int(getenvInt64("COLUMNS"))
+	}
+
+	if width < 40 {
+		width = 80
+	}
+
+	width -= 3
+	if width > 100 {
+		// any wider than this and it gets hard to read
+		width = 100
+	}
+
+	return width
```

Comment:
Not sure why we do this, we're essentially saying you can have between 80 and 100 characters on screen. Even if the user-supplied less we force to 80 or more we say 100, it seems like a lot of effort to say, no.

Reply:
This was from snap info output.

Reply:
no more calculating width.  re-designed to print via yaml.  slices of strings and columns have been converted to string in the yaml struct for readability.